### ISSUE TITLE
Make source and target transport more flexible

### ISSUE THREAD

**Issue by chris-rock** [jnordberg/nsync#2]
Currently `nsync` is fixed to FSTransport and one transport that can differ. It does not make sense to fix it be cause we could allow arbitrary combinations. (https://github.com/jnordberg/nsync/blob/master/src/cli/index.coffee#L260)

I was thinking a while but it is quite difficult to achieve via cli only. The sample 
`nsync sftp --host myserver.com --destination /var/www/my-project ~/my-project`
could also be written as:

``` json
{
    concurrency: '',
    destructive: '',
    source : {
        type: 'fs',
        path: '~/my-project'
    },
    destination: {
        type: 'sftp,
        host: 'myserver.com',
        path: '/var/www/my-project'
    }
}
```

This would allow us to define copy from sftp to sftp, and from s3 to s3. What do you think?

**Comment by jnordberg** [comment]
Yeah, my initial thought was that the cli tool will be fs -> transport and if you want to sync between arbitrary transports you would write a small script using the nsync library

**Comment by chris-rock** [comment]
Sound good, but what is with transport -> fs? As with scp multiple directions are required...

**Comment by jnordberg** [comment]
How about a `--reverse` option?

**Comment by chris-rock** [comment]
that would work for the start I think. I still need to copy data from sftp to s3, but I changed the code in nsync a little bit and got it working. The abstraction of the transport is already quite good. Therefore lets start with a reverse option.

**Comment by jnordberg** [comment]
What did you need to change to make that work? it _should_ work just using the library and passing in the two transports

**Comment by chris-rock** [comment]
I wanted to start this from cli, therefore I quick hacked some changes into the cli to get this working, but is not generalized yet. I was thinking about -stype -soption1 -soption2 and -ttype -toption1 for source and target. But I am not sure if this is a good idea. Anyway with the cli and the lib separated, I can write the tool I need. Sounds good to me. The only thing I stumbled upon is the sync of deletions. e.g. I delete a file locally after the first run, but this is not deleted on the transport.

**Comment by jnordberg** [comment]
Ah ok, yeah i think having that configurability in the cli is too complex. You have to enable the destructive option to make it delete stuff, i'm going to document the options better soon :)

**Comment by chris-rock** [comment]
cool. great project.
